### DISCUSSION TITLE
HTML input datalist with looping and binding.

### DISCUSSION THREAD

**Discussion by simmo9000** [hbroek/reken#22] [Q&A] [answered]
I want to add an HTML drop-down input using a datalist as per [https://developer.mozilla.org/en-US/docs/Web/HTML/Element/datalist](url) 

There is nothing in the reference manual, but something like the below is what I want to achieve.

```

    <div data-if="C" data-calc="updateRecord()">
        <datalist id="ids" data-for="c:C">
            <option data-text="${c.item.label}"></option>
        </datalist>
        <input list="ids" data-value="selected" data-rest="result:${SERVER}?id=${current}" data-action="doGet(c.item.id);" data-rest-options="getRecord">
    </div>

```

This is obviously not going to work because ```c``` which is required in ```data-action="do.Get(c.item.id);"``` is outside of the ```datalist``` for loop. 

If I try this the ```data-rest``` event does not fire as I believe the event is from the ```input```, not ```datalist```

```

<div data-if="C" data-calc="updateRecord()">
        <datalist id="ids" data-for="c:C" data-rest="result:${SERVER}?id=${current}" data-action="doGet(c.item.id);" data-rest-options="getRecord">
            <option data-text="${c.item.label}"></option>
        </datalist>
        <input list="ids" data-value="selected" >
    </div>

```

The following creates an HTML table that triggers the functionality successfully but is not a drop-down.

```

    <table data-if="C">
        <tbody data-for="c:C">
        <tr>
            <td data-text="${c.item.label}"></td>
            <td class="clickable" data-rest="result:${SERVER}?id=${current}" data-action="doGet(c.item.id);" data-rest-options="getRecord" data-text="${c.item.id}"></td>
        </tr>
        </tbody>
    </table>

```

I know I can do this with javascript but it would be neat to do it with Reken.

Any help or suggestions would be appreciated, 
Thanks, Simmo.

**Comment by hbroek** [comment, marked as answer]
HI Simmo,

Thanks for reaching out. If I understand correctly you want a variable `result` populated with the data from the rest call to `SERVER` when the value of the key url parameter changes in the input control. The input control also has a datalist associated with predefined options from which the user can select. I simplified your example and capture the input value `selected` with the input control using the `data-value` attribute. `selected` is then used is the input to the `data-rest` url, which **automatically** refetches when the url changes, in the case by the updated `selected` var. Therefore there is no need for the `data-action` attribute on the input control. 

Note my rest call goes nowhere, but you can see in the network developer tab in chrome/Safari that the network call gets executed every time the `selected` var gets updated. Also for debugging purposes, I added an output control that shows the newly selected value. I moved the `data-rest` attribute to that control as well so I could put them behind a data-if for the start use case when there is no key picked yet. You probably want to move it to the control that displays the contents of `result`.

```html
<html>
<body>
  <div data-if="C">
    <datalist id="ids" data-for="c:C">
        <option data-text="${c.item.label}"></option>
    </datalist>
    <label>Enter/Select key:
      <input list="ids" data-value="selected">
    </label>
    <label data-if="selected!=''" data-rest="result:${SERVER}?id=${selected}">Key entered:
      <output data-text="${selected}"</output>
    </label>
  </div>
</body>
<script>
  let C = [
    {label: "key1"},
    {label: "key2"},
    {label: "key3"},
    {label: "key4"},
  ]
  let selected = ''
  const SERVER = "/json"
  let result
</script>
<script src="reken.js"></script>
</html>
```

I hope this addresses your use case. Note personally more a fan of the easier to use select control for picking values over the input+datalist unless the list is long and you want to enable the user to filter it down.


Best
Henry

**Comment by hbroek** [comment]
Hi @simmo9000,

I added a new demo showing the use of datalist in combination with an input control:

[Datalist demo](https://reken.dev/demo/datalist.html)

Best
Henry

**Comment by simmo9000** [comment]
Thanks Henry,

I appreciate your advice.
This solution gives an easy method of filtering and other manipulation on the input element as a select.  

I really like the easy and simplicity of Reken.  It's great stuff!

Cheers,
Simmo

**Comment by hbroek** [comment]
Hi Simmo,

Thanks, great to hear. Would love to see what you have built with Reken. Planning to add a showcase of apps built with Reken.

Btw can I use your quote for the reken.dev website: It would like something like this:

"I really like the ease and simplicity of Reken. It's great stuff!"
Simmo9000

Thanks
Henry

**Reply by simmo9000** [reply]
Cheers Henry, 
Please feel free to use my quote!
